Utilization of NDPP data for Tallying Estimators with increased efficiency [Much more readable history]

nelsonag opened the pull request:
NOTE: This PR is a continuation of #309 though with a single-commit history so the changes are much clearer as opposed to the 1-1.5 year code development represented in the previous branch.

This pull request gives OpenMC the ability to read and utilize the data necessary to utilize what i'm calling the "improved method" of tallying data with outgoing energy or energy-angle distributions. Right now this only means the tallying of the scattering moment matrices and the fission energy spectra.
Before getting in to the code specifics, its probably useful to discuss (in a very condensed form) why this is needed and how well it works.

# The Method

So if you want to use monte carlo to tally MGXS libraries for a deterministic solver with all data produced to a certain uncertainty level, then most of the computation time will be spent waiting for the scattering moments. They are the last to "converge" because 1) the incoming and outgoing energy filters subdivide the region of phase space, decreasing the total number of tally hits per bin, 2) as the moment order increases, the legendre polynomial becomes more "oscillatory" increasing the apparent variance, and 3) these type of tallies are not conducive to track-length tallying since outgoing collision information is necessary.

To combat this, you would want to use a deterministic scoring technique: that is, given a collision the probability distribution for the outgoing energy and change in angle is provided in the ACE data; tally that complete PDF instead of an individual sampled outgoing point.
Now, that distribution absolutely can not be calculated on the fly at each tally event. The computations involved are way too cumbersome. Instead a pre-processor is needed which takes the ACE data and outputs these probability distribution functions already integrated over outgoing energy and the legendre functional expansion of mu performed.
The pre-processor I wrote for this is called NDPP (http://ndpp.github.io/ source: https://github.com/ndpp/ndpp).

So, how does it perform? Well its not perfect - the tallies obtained using the NDPP data can not ever be identical to what would be obtained using normal analog MC integration (our current scatter-pn type scores). This is both because of truncation error (I use numerical integration techniques in NDPP) as well as interpolation error. In general though the errors are on the order of 0.1 to 0.2%.

But the tallying efficiency is significantly improved as is shown in the Figure below. This figure shows the eigenvalue obtained when generating a P2 MGXS library at each of the number of histories (x-axis) and then pumping this library through MPACT, an MOC code at michigan. Red is using the current nuscatter-pn type tallies. Blue is using the NDPP-based data and track-length estimators. 
Notice that the NDPP-based data significantly reduces the "eigenvalue swing", and (not shown) it does significantly reduce the overall uncertainty on each indvidual tally as you would expect.  Unfortunately it does incur ~15pcm error due to the numerical and interpolation error I mentioned above.
![pcm_p2](https://cloud.githubusercontent.com/assets/1037107/4034592/177e7344-2c87-11e4-8227-28bc96d4d540.png)

Aside from the 15pcm bias, there is one additional problem: tallying the NDPP data is slow. The NDPP tallies were 85% slower than the analog. This could be due to my programming, but also is because the code has to do more work at each collision simply to get the relevant data, interpolate it, and then sum it to the tallies.[ I can significantly improve this if I make the assumption that the user will only want material-wise macroscopic x/s, but I am not willing to make that assumption yet.]

So, if we scale the x-axis of the above figure by run time, then we obtain the figure below (man I need to change that scale).
This shows that despite this slowdown, the NDPP-based data still gets to within 10pcm (the black lines) of its final 1 billion history value (and then stays there) at two hours while the analog method requires ten hours, yielding a factor of 5 improvement. Any speed-up to the NDPP code will directly improve this.
![pcm_p2_time](https://cloud.githubusercontent.com/assets/1037107/4034739/5279ce98-2c88-11e4-9f77-7c9d5ce2e533.png)

Ok, so thats the bare minimum I can give (keeping this PR short) to say why this would be a nice feature. I can of course answer any questions you have about the method here, by e-mail, or in the developers google group.

# The Code 

How did I integrate this in to OpenMC? I created versions of scatter-n/pn/yn and nuscatter-n/pn/yn which use the NDPP data instead. These are simply called ndpp-*.  In addition, I added ndpp-chi and ndpp-chi-p (total fission spectra and prompt fission spectra). These arent discussed much above since the benefit isnt as big as it is for scattering moments but could still definitely be useful.

So before the user can access these scores they have to first run NDPP and produce data for a given set of nuclides and energy group structure.  This will output a bunch of libraries (one for each zaid.kT pair) and an `ndpp_lib.xml` file which is similar to OpenMC's `cross_sections.xml` file but describes the NDPP-produced libraries.

Once this data is generated the user simply has to request one of the ndpp-* scores in a tally and also specify the location of `ndpp_lib.xml` in the `<ndpp_library>` element of `tallies.xml`. The user is free to use analog or tracklength estimators for all these types. After that, its off to the races.

Here are some specifics to help you in your review:
1. I created tests for all the new scores. They require NDPP data, and so I created some sample data in openmc/tests/NDPP_data for 1001, 8016, 92235, and HH2O with 2 groups using the room temperature NNDC data. I purposefully created this in the ASCII format so you could see what the data looks like if you want.
2. When the user requests a NDPP scattering score, they must include energy and energyout filters with bins matching the group structure of the NDPP libraries. The energy filter must be declared immediately before energyout. Also, the scattering order must be less than or equal to that in the NDPP library. If any of these fail, appropriate error messages will be raised. The energy filter must be right before energyout so that the tallies can be efficiently summed.
3. The Chi scores only require an energyout filter, the user can have energy if they please; the order doesnt matter here.
4. Right now only one energy group structure (one `ndpp_lib.xml` file and associated ata) can be used by OpenMC per run. In practice this shouldnt matter, but it could be improved upon in the future.
5. The random number stream is not affected by the presence of the NDPP tallies and so the neutron simulation is unchanged by the presence of NDPP tallies.
6. I didnt mention this with the methods section, but NDPP produces the data such that the elastic and total inelastic data are separate. This is because the elastic data is temperature dependent anad the inelastic data is independent of kT.  OpenMC takes advantage of this by throwing out the redundant temperature independent NDPP data which could exist say if 92235.71c and 92235.72c both existed in a model.
7. I did not modify CMFD to take advantage of the NDPP data. This was because:
 a. When item 4 is addressed this could become trivial. I think of item 4 as a pre-requisite since one probably would prefer to do like 47-group tallies, but only do a 2-grp CMFD to get there.
 b. It may not be that big of a deal, I *think* that CMFD's convergence on an eigenvalue depends more on the current since D-hat is adjusted from its normal 1/(3sigtr) value to preserve currents.

nelsonag commented:
Note that I am currently re-running the latest version of NDPP to re-generate the most up-to-date data possible to be placed in the tests/NDPP_Data directory and used for the ndpp-specific tests.  That will be the next commit.

nelsonag commented:
As far as I can tell this should work, but I have some seemingly unrelated tests failing now like test_score_fission.  I will be tracking this down.

nelsonag commented:
Alright, everything should be set to go.  The previous issue (tests like test_score_fission failing) was actually due to my needing to get the NNDC data again, for some reason.  Now every test passes.

wbinventor commented on docs/source/usersguide/input.rst:
```diff
@@ -1099,29 +1099,38 @@ The ``<tally>`` element accepts the following sub-elements:
+    "scatter-YN", "nu-scatter-N", "nu-scatter-PN", "nu-scatter-YN",
+    "ndpp-scatter-N", "ndpp-scatter-PN", "ndpp-scatter-YN", "ndpp-nu-scatter-N",
+    "ndpp-nu-scatter-PN", "ndpp-nu-scatter-YN", "flux-YN", "total-YN",
+    "current", and "events".  These corresponding to the following physical
```
This list is missing "ndpp-chi," "ndpp-chi-p" and "ndpp-chi-d."

smharper replied:
leave out "ndpp-chi-d" since it's not supported

wbinventor commented on docs/source/usersguide/input.rst:
```diff
@@ -1162,6 +1175,20 @@ The ``<tally>`` element accepts the following sub-elements:
       scattering is scored vice simply the scattering rate. This accounts for
       multiplicity from (n,2n), (n,3n), and (n,4n) reactions.
 
+    :ndpp-scatter-N, ndpp-scatter-PN, ndpp-scatter-YN, ndpp-nu-scatter-N,
```
Sphinx doesn't compile this correctly, perhaps due to the line break.

wbinventor commented:
The documentation for this PR mentions that the "ndpp-*" tally filters must be used with "energy" and "energyout" filters specified in that order. In addition, the energy bins for both filters must be identical. Based on my limited experience trying to use this functionality, it seems that it would be very useful to relax each of these constraints, particularly the second one.

For example, removing these constraints would be particularly useful when computing transport-corrected total cross-sections which require a "total" and "scatter-1" scores. At the moment, OpenMC only supports analog estimators for "scatter-N" filter types, but with NDPP, this is no longer the case. However, the "scatter-1" score in this case (implicitly) needs an "energyout" filter with bins [0, infty] (*e.g*, ``bins="0 20"``), but this will not match the "energy" filter's bins unless one is only computing one group cross-sections. Instead, the user must specify an "energyout" filter with the same bins, then do downstream data processing to reduce the tallies from each "energyout" filter bin to compute the same result. In the end, this requires 1) more tally data storage and 2) a greater burden on the user.

In addition, it would be nice to remov the second constraint mentioned above for scenarios such as this one by implicitly defining a default "energyout" filter with bins [0, infty] to encompass all of energy space if none is specified. 

My guess is that an attempt to address these concerns would run into point 4 mentioned by @nelsonag in the description for this PR - namely that only a single energy group structure may be used for a particular OpenMC simulation with NDPP tally scores.

@nelsonag - what are your thoughts on this issue? Considering that this PR is specifically intended to make it easier (or at least faster) to compute MGXS with OpenMC, wouldn't it make sense to make it easier to compute transport-corrected total cross-sections by relaxing these constraints? Or am I missing something altogether here?

nelsonag commented:
There are two separate issues youve brought up that I'd like to keep
separate: 1) having energy and energyout in that order, and 2) making
energyout optional and its lack of presence implying [0,infty).

1) if the user does want specific energyout filters (if they do not use
item 2) then I am able to make the tallying routines significantly faster
since I know exactly the layout of the filters given the starting filter
index.  This allows me to avoid doing work for each tally t o figure out
where I am summing data to.  Next, this also means much better caching
performance as you know.  Since I think this improvement is worth more than
the burden it may cause to a very few users who have post processing
scripts set up that assume a different (and non-adjusting) set up, I'm
willing to keep this constraint.  From your note it doesn't seem like you
are too hard up on this item either.

2) Hmm this is interesting. Until the ability to load multiple ndpp data
libraries is implemented, the only way I can do this is by writing 2 new
tallying routines (for scatter-n and scatter-on) which basically sum up the
ndpp data over outgoing energy groups and score to the given tally filter
location.  I would also need some logic in there to flag when these
routines are needed of course.  I am on the fence with this as it means we
are only supporting one transport cross section method (outscatter) instead
of leaving it generic so the user can generate the raw data needed and do
it with only post-processing.  I'd like to have more discussion on this and
get anyone else's thoughts too.

Adam
On Feb 21, 2015 11:59 PM, "Will Boyd" <notifications@github.com> wrote:

> The documentation for this PR mentions that the "ndpp-*" tally filters
> must be used with "energy" and "energyout" filters specified in that order.
> In addition, the energy bins for both filters must be identical. Based on
> my limited experience trying to use this functionality, it seems that it
> would be very useful to relax each of these constraints, particularly the
> second one.
>
> For example, removing these constraints would be particularly useful when
> computing transport-corrected total cross-sections which require a "total"
> and "scatter-1" scores. At the moment, OpenMC only supports analog
> estimators for "scatter-N" filter types, but with NDPP, this is no longer
> the case. However, the "scatter-1" score in this case (implicitly) needs an
> "energyout" filter with bins 0, infty, but this will not match the
> "energy" filter's bins unless one is only computing one group
> cross-sections. Instead, the user must specify an "energyout" filter with
> the same bins, then do downstream data processing to reduce the tallies
> from each "energyout" filter bin to compute the same result. In the end,
> this requires 1) more tally data storage and 2) a greater burden on the
> user.
>
> In addition, it would be nice to remov the second constraint mentioned
> above for scenarios such as this one by implicitly defining a default
> "energyout" filter with bins [0, infty] to encompass all of energy space if
> none is specified.
>
> My guess is that an attempt to address these concerns would run into point
> 4 mentioned by @nelsonag <https://github.com/nelsonag> in the description
> for this PR - namely that only a single energy group structure may be used
> for a particular OpenMC simulation with NDPP tally scores.
>
> @nelsonag <https://github.com/nelsonag> - what are your thoughts on this
> issue? Considering that this PR is specifically intended to make it easier
> (or at least faster) to compute MGXS with OpenMC, wouldn't it make sense to
> make it easier to compute transport-corrected total cross-sections by
> relaxing these constraints? Or am I missing something altogether here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/361#issuecomment-75419908>.
>

wbinventor commented:
I agree with you that the first issue is not really a problem. After re-reading my post today, I see that I accidentally implied that the "energy" and "energyout" ordering was important to me from a post-processing standpoint, but in reality, it is only issue 2 that I think needs to be addressed.

Perhaps one option for addressing the second issue is to include one group data along with the multi-group data with every NDPP library that is generated for use in cases when an energyout filter is not specified. Of course this would require both a modification to NDPP as well as to the code in this PR. 

I recognize that with some data processing, users can get around issue 2, but it does require **much** more data to be generated and stored in the statepoint files when tallying for a MGXS library. With analog tallies, the size of the tallied data (very roughly) scales with the number of energy groups squared for the scattering matrix. With this implementation for NDPP, the size of the data doubles for the outscatter transport cross-section. This likely isn't burdensome for simple pin cell problems, but would become an issue for larger problems (even 2D lattice physics problems with ~1000 tally zones).

nelsonag commented:
Well I can generate it a priori by openmc via a sum over outgoing groups so
no need to have ndpp produce it as well. I guess I'm just wondering if we
were going to produce a transport correction with an MC code would it be
outscatter? Ideally we would use a harmonic weighted in-scatter type system
but that doesn't always produce positive xs at lower energies. The
outscatter is only desirable for getting around that issue.

@bhermanmit do you have any opinions?

I agree with you that the first issue is not really a problem. After
re-reading my post today, I see that I accidentally implied that the
"energy" and "energyout" ordering was important to me from a
post-processing standpoint, but in reality, it is only issue 2 that I think
needs to be addressed.

Perhaps one option for addressing the second issue is to include one group
data along with the multi-group data with every NDPP library that is
generated for use in cases when an energyout filter is not specified. Of
course this would require both a modification to NDPP as well as to the
code in this PR.

I recognize that with some data processing, users can get around issue 2,
but it does require *much* more data to be generated and stored in the
statepoint files when tallying for a MGXS library. With analog tallies, the
size of the tallied data (very roughly) scales with the number of energy
groups squared for the scattering matrix. With this implementation for
NDPP, the size of the data doubles for the outscatter transport
cross-section. This likely isn't burdensome for simple pin cell problems,
but would become an issue for larger problems (even 2D lattice physics
problems with ~1000 tally zones).

—
Reply to this email directly or view it on GitHub
<https://github.com/mit-crpg/openmc/pull/361#issuecomment-75467042>.

smharper commented on src/tally.F90:
```diff
@@ -383,6 +385,124 @@ score = wgt * mu ! avoid function call overhead
+              matching_bins(t % find_filter(FILTER_ENERGYIN)), &
+              score_index, filter_index - &
+              matching_bins(t % find_filter(FILTER_ENERGYOUT)) + 1, &
+              t % moment_order(j), last_wgt, .True., p % last_E, t % results, .True.)
```
We should use the post-collision weight rather than the pre-collision weight for nu-scatter, right?

nelsonag replied:
Good question. Since NDPP combines all the scattering reaction channels in
to one set of data, it already factors in the reaction multiplicity during
this combination step.  So using post-collision weight here would be double
counting the multiplicity.


On Tue, Jun 23, 2015 at 1:19 AM, Sterling Harper <notifications@github.com>
wrote:

> In src/tally.F90
> <https://github.com/mit-crpg/openmc/pull/361#discussion_r33010215>:
>
> > +            end if
> > +
> > +            j = j + (t % moment_order(j) + 1)**2 - 1
> > +            cycle SCORE_LOOP
> > +
> > +          case (SCORE_NDPP_NU_SCATT_N)
> > +            ! Skip any event where the particle didn't scatter
> > +            if (p % event /= EVENT_SCATTER) cycle SCORE_LOOP
> > +
> > +            ! For the case of analog NDPP nu-scatter-n tallying, filter_index needs
> > +            ! to be adjusted to point to the first energyout filter
> > +            call tally_ndpp_n(p % event_nuclide, &
> > +              matching_bins(t % find_filter(FILTER_ENERGYIN)), &
> > +              score_index, filter_index - &
> > +              matching_bins(t % find_filter(FILTER_ENERGYOUT)) + 1, &
> > +              t % moment_order(j), last_wgt, .True., p % last_E, t % results, .True.)
>
> We should use the post-collision weight rather than the pre-collision
> weight for nu-scatter, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/361/files#r33010215>.
>

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+!===============================================================================
+! READ_NDPP_XML reads information from a cross_sections.xml file. This
+! file contains a listing of the ACE cross sections that may be used.
+!===============================================================================
```
Looks like this wasn't updated after a copy-paste

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+      call get_node_array(doc, "energy_bins", ndpp_energy_bins)
+    end if
+
+    ! Get if nuscatter is present
```
A little picky, but I'd prefer "Check" to "Get" here and the below comments.

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+      SCORE_LOOP: do k = 1, t % n_user_score_bins
+        j = j + 1
+        select case (t % score_bins(j))
+          case (SCORE_SCATTER_PN)
```
Our convention is to indent the `case` statements at the same level as `select case`

smharper replied:
There is a lot of repeated code in these different cases.  I would recommend making one `select case` block that checks the order and filters, and then another `select case` block that adjusts `j`

smharper replied:
Is there any particular reason you specify a case for `SCORE_SCATTER_PN`, but none of the other non-NDPP scores?  Honestly I don't think the extra code for the non-NDPP cases is worth the computation time that it saves.  I think you should just remove this case.

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+            ! can use the Fortran intrinsic ALL to check the actual values
+            if (t % filters(i_filter) % n_bins /= ndpp_groups) then
+              call fatal_error("Number of groups in NDPP Library do " // &
+                   "not match that requested in tally!")
```
Please add the tally id to the error messages

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+    if (.not. file_exists) then
+       ! Could not find ndpp_lib.xml file
+       call fatal_error("NDPP Library XML file '" // trim(ndpp_lib) // &
+            "' does not exist!")
```
This is indented three spaces (should be two)

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+       ! directory in which the ndpp_lib.xml file resides
+       i = index(ndpp_lib, "/", BACK=.true.)
+       directory = ndpp_lib(1:i)
+    end if
```
There's a lot of three space indentation.  Maybe you should write a quick script to check for it.

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+       ! copy a number of attributes
+       call get_node_value(node_ndpp, "name", listing % name)
+       if (check_for_node(node_ndpp, "alias")) &
+         call get_node_value(node_ndpp, "alias", listing % alias)
```
Indent continued lines five spaces

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+          listing % type = ACE_THERMAL
+       end if
+
+       ! set filetype, record length, and number of entries
```
remove record length and number of entries from this comment

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+          listing % metastable = .false.
+       else
+          listing % metastable = .true.
+       end if
```
Some of the code down here is indented four rather than two

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+    logical :: get_chi_t     ! Flag for whether or not to get total chi data
+    logical :: get_chi_p     ! Flag for whether or not to get prompt chi data
+    logical :: get_chi_d     ! Flag for whether or not to get delayed chi data
+    integer :: scatt_order   ! Number of moments requested in tallies for scatter
```
This comment is one character over 80 columns

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+    ! Read the ndpp_lib.xml file
+    call read_ndpp_xml(scatt_type, ndpp_groups)
+
+    ! Determine which data is required to be stored as well as the maximum orders
```
81 cols here as well

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+    logical, intent(out) :: get_chi_t   ! Flag for whether or not to get total chi data
+    logical, intent(out) :: get_chi_p   ! Flag for whether or not to get prompt chi data
+    logical, intent(out) :: get_chi_d   ! Flag for whether or not to get delayed chi data
+    integer, intent(out) :: scatt_order ! Number of moments requested in tallies for scatter
```
Please get these guys down to 80 cols as well.  I think our convention for this is to indent multiple lines of comments so that all the exclamations line up (even these little inline guys at the top of subprograms)

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+      SCORE_LOOP: do k = 1, t % n_user_score_bins
+        j = j + 1
+        select case (t % score_bins(j))
+          case (SCORE_NDPP_SCATT_N, SCORE_NDPP_SCATT_PN, SCORE_NDPP_SCATT_YN)
```
See earlier comment about `select case` indention

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+            else if (t % score_bins(j) == SCORE_NDPP_SCATT_YN) then
+              j = j + (t % moment_order(j) + 1)**2 - 1
+              cycle SCORE_LOOP
+            end if
```
Again, I think the compute time you save is likely not worth the extra code (mostly because someone might try to modify it one day and mess it up, not because of lines-of-code statistics or something like that)

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+    integer, intent(in) :: scatt_order ! Number of moments requested in tallies
+    logical, intent(in) :: is_nuc      ! Is our data a nuc or an sab?
+
+    integer       :: in = 7        ! file unit
```
Replace 7 with a parameter variable, please

smharper replied:
I see that you copied that from ace.F90.  I'd appreciate it if you replaced the `in` variable in ace.F90 as well.  I guess they should both be using a parameter variable from constants.F90 named `DATA_FILE_HANDLE` or something like that...

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+    real(8), allocatable :: temp_outgoing(:,:) ! Temporary storage of scatt data
+    integer       :: NP            ! Number of precursors groups
+
+    ! determine path, record length, and location of table
```
Replace record length with file type in the comment

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+        end if
+      end do
+
+      ! The remainder only apply to nuclides (inelastic, nu-inelastic and chi data)
```
80 columns please

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+              ! And copy in to this % el
+              allocate(this % inel(iE) % outgoing(scatt_order, gmin : gmax))
+              this % inel(iE) % outgoing(:, gmin : gmax) = &
+                temp_outgoing(1 : scatt_order, gmin : gmax)
```
5 spaces indent here

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+            if ((gmin > 0) .and. (gmax > 0)) then
+              ! Then we can allocate the space. Do it to lib_order
+              ! since this is the largest order requested in the tallies.
+              ! Since we only need to store up to the maximum, we also need to have
```
80 columns

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+                ! since this is the largest order requested in the tallies.
+                ! Since we only need to store up to the maximum, we also need to have
+                ! an array for reading the file which we can later truncate to fit
+                ! in to this % nuinel(iE) % outgoing.
```
80 columns

smharper replied:
This comment still needs to be addressed

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+                allocate(this % nuinel(iE) % outgoing(scatt_order, &
+                  gmin : gmax))
+                this % nuinel(iE) % outgoing(:, gmin : gmax) = &
+                  temp_outgoing(1 : scatt_order, gmin : gmax)
```
5 spaces indent

smharper replied:
This one does too

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+              end if
+            end do
+          end if
+        end if
```
This code is nested pretty deep.  It would help readability to put some labels on some of the outermost loops.

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+      ! the right data set), and the other meta information
+      allocate(energy_bins(NG + 1))
+      read(UNIT=in) energy_bins, scatt_type, lib_order, &
+        nuscatter, chi_present, mu_bins, thin_tol
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+          ! And copy in to this % el
+          allocate(this % el(iE) % outgoing(scatt_order, gmin : gmax))
+          this % el(iE) % outgoing(:, gmin : gmax) = &
+            temp_outgoing(1 : scatt_order, gmin : gmax)
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+              ! since this is the largest order requested in the tallies.
+              ! Since we only need to store up to the maximum, we also need to have
+              ! an array for reading the file which we can later truncate to fit
+              ! in to this % inel(iE) % outgoing.
```
80 columns.  Gosh, we really should try to get our CI to check these pedantic little formatting things.  My UROP project senses are tingling.

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+              ! And copy in to this % inel
+              allocate(this % inel(iE) % outgoing(scatt_order, gmin : gmax))
+              this % inel(iE) % outgoing(:, gmin : gmax) = &
+                temp_outgoing(1 : scatt_order, gmin : gmax)
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+        end if
+
+        ! Get chi data, if provided
+        if (chi_present == 1) then
```
Make `chi_present` a bool

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+      ! =======================================================================
+      ! READ NDPP DATA IN HDF5 FORMAT
+
+      ! NOT YET IMPLEMENTED
```
Make this spit out an error.

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+            data % inel_Ein_srch => ref % inel_Ein_srch
+          end if
+
+          ! Ok, good work team, now mark skip as true so we dont do it again
```
Cheeky banter should be a requirement in all new code > 1000 lines :)

nelsonag replied:
I agree :-)

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,622 @@
+  end subroutine read_ndpp_data
+
+!===============================================================================
+! REMOVE_TEMPERATURE_INDEPENDENT_DATA sweeps through the NDPP Nuclide data looking
```
80 cols

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+        ! get gmin and gmax
+        read(UNIT=in, FMT='(I20,I20)') gmin, gmax
+
+        if ((gmin > 0) .and. (gmax > 0)) then
```
When does this conditional fail?  Do we check somewhere to make sure that this data was allocated if we need it?

nelsonag replied:
This conditional fails when the NDPP data has no distribution for an incoming energy point.  This can theoretically occur (though it is not expected) when, for example, the only inelastic reaction channel present at an energy is described by a law that NDPP does not cover yet (because its crazy rare and I havent seen it yet), or sometimes if the ACE data is not self consistent.

During the tallying process (specifically within generate_ndpp_distrib_*n) this is checked.  It needs to be checked for other purposes besides this rare case, such as if the next data point is not allocated (because you are exactly at the upper bound) so there sadly isnt much time savings to be had by just checking up front.

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+        micro_xs(i_nuclide) % absorption)
+    end if
+
+    ! Add the contribution from the elastic score
```
This includes the contribution from both elastic and inelastic, right?

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1083 @@
+        results(i_score: i_score + num_lm - 1, g_filter) % value = &
+          results(i_score: i_score + num_lm - 1, g_filter) % value + &
+          ndpp_outgoing(thread_id, l + 1, g) * norm * calc_rn(l, uvw)
+!$omp end critical
```
Can we not use the `atomic` pragma here?  (Forgive my ignorance of OpenMP)

nelsonag replied:
Almost, `atomic` is reserved for instructions which processors have specific instructions for and thus you usually only see them for very simple changes like `i = i + 1`.  Critical is for everything else.  Since this block has array slicing, multiple multiplications and additions, I kept it critical.

However.... in making sure I was remembering correctly, I did a google search on your question.  Well, I found out that `omp critical` sections should be named, otherwise they are all treated as the same critical question.  I did not know that.  So, now all of these critical sections are named per my last commit which I will make once finished addressing your comments. I will also create a PR to do the same for the rest of OpenMC.

smharper commented:
Add some error checking that will alert the user if their filters don't match the NDPP data.  Right now, if I change the filters in the tests, they will either get the same results or fail in a binary search.

smharper commented:
I think we should have some tests for scattering orders beyond first.

smharper commented:
This is a cool PR @nelsonag :).  I think I definitely want to get it merged before I keep fooling with tallies.  Looking at this has already changed the design I had in my mind for the tally refactor, and if we play our cards right, a refactor might be able to give NDPP a speed-up.

nelsonag commented:
Thanks @smharper! I will get these comments incorporated tonight or tomorrow night.  

I have also started working on a speed improvement to pre-combine the ndpp data in to one set per material. This should significantly improve performance of NDPP tallying at the potential cost of memory usage.  Of course this can't be used for tallies with nuclide filters.  Anyways this will be in a separate PR.

nelsonag commented:
> Add some error checking that will alert the user if their filters don't match the NDPP data. Right now, if I change the filters in the tests, they will either get the same results or fail in a binary search.

Oops.  Good catch; the problem was I was using `all` in the test conditional I had instead of `any` so now the check I had intended to perform should be working.

nelsonag commented:
Alright @smharper, the only remaining comment I have to address is to expand the number of scattering orders represented by the tests.  I have to run NDPP for that and I will just run that before going to bed.  (No it doesnt take that long :-) I want to use the computer for a little more of my waking hours before giving it a large workload)

nelsonag commented:
Oh no. Seemingly random travis build errors again.  This past time the travis_install.sh failed during the python3 portion.  Sigh.  Will attack that on the next go-round.

smharper commented:
There's a way to rerun Travis builds.  It's a little circular arrow symbol on the right side of the screen.  It's rerunning now.

paulromano commented:
@nelsonag I apologize I haven't paid more attention to your PR thus far. With the recent interest from @wbinventor and @smharper, I figured I should also make an effort to review so we can get this merged in.

A few basic planning questions:
- Do you want NDPP itself to remain separate? Or would you rather it become part of OpenMC itself? Another option would be to have it as a submodule like FoX currently is.
- Do you plan on maintaining this functionality into the foreseeable future?
- What specific plans do you have for improvements in NDPP going forward?

The test failures look real right now (all NDPP-related tests failing and all others passing).

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,604 @@
+    ! Determine which data is required to be stored as well as the maximum
+    ! orders for scatt and nuscatt
+    call which_data(scatt_type, get_scatt, get_nuscatt, get_chi_t, get_chi_p, &
+                    get_chi_d, scatt_order)
```
Hey @paulromano, I like this style of indentation where a continued line is indented past the five spaces to match it's context.  Would you be opposed to explicitly allowing this in the developers guide?

paulromano replied:
Not opposed at all!

smharper commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,604 @@
+         call get_node_value(doc, "scatt_type", scatt_type)
+    if (scatt_type /= SCATT_TYPE_LEGENDRE) then
+      call fatal_error("Invalid Scattering Type represented in NDPP " // &
+           "data. Rerun NDPP with the Legendre scattering type set.")
```
You don't need the `//` operator to concatenate these strings.  Just put an `&` at the end of each line (without closing the quotes) and another `&` at the beginning of the next line (without reopening the quotes).  Check out input_xml.F90 for examples of this.

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+          ! And copy in to this % el
+          allocate(this % el(iE) % outgoing(scatt_order, gmin : gmax))
+          this % el(iE) % outgoing(:, gmin : gmax) = &
+            temp_outgoing(1 : scatt_order, gmin : gmax)
```
5 spaces indent, please

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+    ! Now apply sigS if we are in tracklength mode
+    if (.not. is_analog) then
+      norm = norm * (micro_xs(i_nuclide) % total - &
+        micro_xs(i_nuclide) % absorption)
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+!$omp atomic
+      results(score_index, g_filter) % value = &
+        results(score_index, g_filter) % value + &
+        ndpp_outgoing(thread_id, l, g) * norm
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+    ! Now apply sigS if we are in tracklength mode
+    if (.not. is_analog) then
+      norm = norm * (micro_xs(i_nuclide) % total - &
+        micro_xs(i_nuclide) % absorption)
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+        i_score = score_index + l - 1
+!$omp atomic
+        results(i_score, g_filter) % value = &
+          results(i_score, g_filter) % value + ndpp_outgoing(thread_id, l, g) * norm
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+    ! Now combine the elastic and inelastic distributions to one set,
+    ! which will be in ndpp_outgoing(thread_id,:,:)
+    call generate_ndpp_distrib_pn(this, nuscatt_flag, gin, t_order, Ein, sigs_el, sigs_inel, &
+                                  norm, gmin, gmax)
```
80 columns

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+    ! Now apply sigS if we are in tracklength mode
+    if (.not. is_analog) then
+      norm = norm * (micro_xs(i_nuclide) % total - &
+        micro_xs(i_nuclide) % absorption)
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+!$omp critical (ndpp_tally_scatt_yn)
+        results(i_score: i_score + num_lm - 1, g_filter) % value = &
+          results(i_score: i_score + num_lm - 1, g_filter) % value + &
+          ndpp_outgoing(thread_id, l + 1, g) * norm * calc_rn(l, uvw)
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+    else
+      i_grid = binary_search(chi_Ein, size(chi_Ein), Ein)
+      f = (Ein - chi_Ein(i_grid)) / &
+        (chi_Ein(i_grid + 1) - chi_Ein(i_grid))
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+      results(score_index, g_filter) % value = &
+        results(score_index, g_filter) % value + &
+        chi(g, i_grid) * one_f + &
+        chi(g, i_grid + 1) * f
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+         nuclides(i_nuclide) % absorption(micro_xs(i_nuclide) % index_grid)) + &
+        (micro_xs(i_nuclide) % interp_factor) * &
+        (nuclides(i_nuclide) % total(micro_xs(i_nuclide) % index_grid + 1) - &
+         nuclides(i_nuclide) % absorption(micro_xs(i_nuclide) % index_grid + 1))
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+    else
+      sigs_el   = micro_xs(i_nuclide) % elastic
+      sigs_inel = micro_xs(i_nuclide) % total - micro_xs(i_nuclide) % absorption - &
+        sigs_el
```
80 columns

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+      el_igrid = binary_search(this % el_Ein(srch_lo: srch_hi), &
+                             srch_hi - srch_lo + 1, Ein) + srch_lo - 1
+      el_f = log(Ein / this % el_Ein(el_igrid)) / &
+        log(this % el_Ein(el_igrid + 1) / this % el_Ein(el_igrid))
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+          inel_igrid = binary_search(this % inel_Ein(srch_lo: srch_hi), &
+                                 srch_hi - srch_lo + 1, Ein) + srch_lo - 1
+          inel_f = log(Ein / this % inel_Ein(inel_igrid)) / &
+            log(this % inel_Ein(inel_igrid + 1) / this % inel_Ein(inel_igrid))
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+      do g = lbound(this % el(el_igrid) % outgoing, dim=2), &
+             ubound(this % el(el_igrid) % outgoing, dim=2)
+        ndpp_outgoing(thread_id, l, g) = el_one_f * &
+          this % el(el_igrid) % outgoing(l, g)
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+             ubound(this % el(el_igrid + 1) % outgoing, dim=2)
+        ndpp_outgoing(thread_id, l, g) = &
+          ndpp_outgoing(thread_id, l, g) + &
+          el_f * this % el(el_igrid + 1) % outgoing(l, g)
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+        do g = lbound(inel(inel_igrid) % outgoing, dim=2), &
+               ubound(inel(inel_igrid) % outgoing, dim=2)
+          ndpp_outgoing(thread_id, l, g) = ndpp_outgoing(thread_id, l, g) + &
+            inel(inel_igrid) % outgoing(l, g) * inel_one_f
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+        do g = lbound(inel(inel_igrid + 1) % outgoing, dim=2), &
+               ubound(inel(inel_igrid + 1) % outgoing, dim=2)
+          ndpp_outgoing(thread_id, l, g) = ndpp_outgoing(thread_id, l, g) + &
+            inel(inel_igrid + 1) % outgoing(l, g) * inel_f
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+      do g = lbound(this % el(el_igrid) % outgoing, dim=2), &
+             ubound(this % el(el_igrid) % outgoing, dim=2)
+        ndpp_outgoing(thread_id, 1: l + 1, g) = el_one_f * &
+          this % el(el_igrid) % outgoing(1: l + 1, g)
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+             ubound(this % el(el_igrid + 1) % outgoing, dim=2)
+        ndpp_outgoing(thread_id, 1: l + 1, g) = &
+          ndpp_outgoing(thread_id, 1: l + 1, g) + &
+          el_f * this % el(el_igrid + 1) % outgoing(1: l + 1, g)
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+               ubound(inel(inel_igrid) % outgoing, dim=2)
+          ndpp_outgoing(thread_id, 1: l + 1, g) = &
+            ndpp_outgoing(thread_id, 1: l + 1, g) + &
+            inel(inel_igrid) % outgoing(1: l + 1, g) * inel_one_f
```
5 spaces indent

smharper commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+               ubound(inel(inel_igrid + 1) % outgoing, dim=2)
+          ndpp_outgoing(thread_id, 1: l + 1, g) = &
+            ndpp_outgoing(thread_id, 1: l + 1, g) + &
+            inel(inel_igrid + 1) % outgoing(1: l + 1, g) * inel_f
```
5 spaces indent

paulromano commented on src/ace.F90:
```diff
@@ -228,13 +228,14 @@ subroutine read_ace_table(i_table, i_listing)
     integer, intent(in) :: i_table   ! index in nuclides/sab_tables
     integer, intent(in) :: i_listing ! index in xs_listings
 
+    integer       :: in = UNIT_DATAFILE ! file unit
```
Rather than create a new hard-coded unit, we should just be using `open(NEWUNIT=...)`.

paulromano commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,604 @@
+    do i_nuclide = 1, n_nuclides_total
+      nuc => nuclides(i_nuclide)
+      i_listing = ndpp_listing_dict % get_key(adjustl(trim(nuc % name)))
+      if (i_listing == DICT_NULL) then
```
A more idiomatic way of doing this would be:
```fortran
if (.not. ndpp_listing_dict % has_key(adjustl(trim(nuc % name)))) then
```
Then you don't have to make `DICT_NULL` public.

paulromano commented on src/input_xml.F90:
```diff
@@ -2152,6 +2152,13 @@ subroutine read_tallies_xml()
+    if (check_for_node(doc, "ndpp_library")) then
+      call get_node_value(doc, "ndpp_library", ndpp_lib)
+    else
+      ndpp_lib = './ndpp_lib.xml'
```
No need for the dot slash.

paulromano commented on src/ndpp_header.F90:
```diff
@@ -0,0 +1,109 @@
+    real(8), pointer :: chi(:,:) => null()           ! Data grid for ndpp chi data
+                                                ! dimensions of chi: (g, Ein)
+    real(8), pointer :: chi_p(:,:) => null()         ! Same for prompt only
+    real(8), pointer :: chi_d(:,:,:) => null()       ! Same, but additional dimension
```
Do all of these components really need to be `pointer`, or can they be `allocatable` instead (much preferred)?

nelsonag replied:
I have made all but the inelastic data as pointers.  I could not move them as discussed in the comment beginning on line 30.

paulromano commented on src/ndpp_header.F90:
```diff
@@ -0,0 +1,109 @@
+  type GrpTransfer
+    real(8), allocatable :: outgoing(:,:) ! Outgoing transfer probabilities
+                                          ! Dimension of (moments, gmin:gmax)
+  end type GrpTransfer
```
I'd prefer the ever-so-slightly more verbose but less ambiguous `GroupTransfer`.

paulromano commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,604 @@
+  subroutine read_ndpp_data()
+    type(Nuclide), pointer    :: nuc => null() ! Current working nuclide
+    type(SAlphaBeta), pointer :: sab => null() ! Current working SAB table
+    type(XsListing), pointer  :: ndpp_listing => null()
```
In sticking with recent changes, locally-declared pointers should not be initialized to null() at declaration since this causes them to have static duration.

paulromano commented on src/ndpp_header.F90:
```diff
@@ -0,0 +1,109 @@
+      procedure :: clear => ndpp_clear ! Clears NDPP structure
+  end type Ndpp
+
+  contains
```
In all other modules, contains appears with no indentation.

paulromano commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,604 @@
+
+  ! ndpp_lib.xml preprocessed data listings and associated data.
+  type(XsListing), allocatable, target, save :: ndpp_listings(:)
+  type(DictCharInt), save                    :: ndpp_listing_dict
```
`SAVE` is implicit for module variables. From the F2008 standard:

*A variable, common block, or procedure pointer declared in the scoping unit of a main program, module, or submodule implicitly has the SAVE attribute, which may be confirmed by explicit specification.*

paulromano commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,604 @@
+                          ! other temperatures of same ZAID
+    logical, allocatable :: skip(:)        ! Data safe to skip (already done)
+    type(Ndpp), pointer  :: data => null() ! Current data to change
+    type(Ndpp), pointer  :: ref  => null() ! Current data to point to
```
Same comment about removing `=> null()`

paulromano commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,604 @@
+    logical, intent(out) :: get_chi_d   ! Flag for if delay chi data is needed
+    integer, intent(out) :: scatt_order ! Number of scatter moments requested
+
+    type(TallyObject), pointer :: t => null()
```
Same comment about removing `=> null()`

paulromano commented on src/ndpp_init.F90:
```diff
@@ -0,0 +1,604 @@
+    type(XsListing), pointer :: listing => null()
+    type(Node), pointer      :: doc => null()
+    type(Node), pointer      :: node_ndpp => null()
+    type(NodeList), pointer  :: node_ndpp_list => null()
```
Same comment about removing `=> null()`

paulromano commented on src/ndpp_ops.F90:
```diff
@@ -0,0 +1,1102 @@
+    integer, intent(in) :: scatt_order ! Number of moments requested in tallies
+    logical, intent(in) :: is_nuc      ! Is our data a nuc or an sab?
+
+    integer       :: in = UNIT_DATAFILE ! file unit
```
Same comment about using NEWUNIT

paulromano commented on tests/readme.rst:
```diff
@@ -55,4 +55,4 @@ can run:
 
 .. _ctest: http://www.cmake.org/cmake/help/v2.8.12/ctest.html
-.. _NNDC:  http://http://www.nndc.bnl.gov/endf/b7.1/acefiles.html
+.. _NNDC:  http://www.nndc.bnl.gov/endf/b7.1/acefiles.html
```
Ooh, nice catch!

paulromano commented:
The new tests for this PR need to be refactored based on #402.

nelsonag commented:
>@nelsonag I apologize I haven't paid more attention to your PR thus far. With the recent interest from @wbinventor and @smharper, I figured I should also make an effort to review so we can get this merged in.

>A few basic planning questions:
•Do you want NDPP itself to remain separate? Or would you rather it become part of OpenMC itself? Another option would be to have it as a submodule like FoX currently is.
•Do you plan on maintaining this functionality into the foreseeable future?
•What specific plans do you have for improvements in NDPP going forward?

>The test failures look real right now (all NDPP-related tests failing and all others passing).

Not a problem @paulromano! You ask some valid questions which I'm glad you have asked now as opposed to as soon as I posted this PR as I think my answers have changed.  

>•Do you want NDPP itself to remain separate? Or would you rather it become part of OpenMC itself? Another option would be to have it as a submodule like FoX currently is.

I think since I have already done the work to split NDPP from OpenMC, we might as well keep it separate (or a submodule which I am fine with).
This answer of course depends heavily on the next, so I suppose I will wait on a final opinion until we decide the next.

>•Do you plan on maintaining this functionality into the foreseeable future?

This is an interesting question with an interesting response.  The direct answer is that I do plan on maintaining this functionality in OpenMC and NDPP.
What makes this interesting is that the past 10 months have taught me that I don't have as much time to devote to this as I intended to when this PR was first submitted.
I don't prefer this to be the case, but sadly this has to be at a hobby level for me.

When this PR was pushed, I had (naively) hoped that the incorporation of the code in the main version would drive increased interest in NDPP which could then increase the odds that someone would make NDPP more than just a one-man show.
We haven't gotten it in to the main version yet but the more skeptical (i.e., post-graduation) version of me thinks this might not be the case, and adding the implicit reliance on the stability and longevity of NDPP into OpenMC via a non-negligible amount of complicated code does represent a risk to our team.

So if you think it is not the right decision to incorporate this large PR in the main codebase and would rather have me maintain a separate fork (exactly as I do now), I 100% understand and support that decision.

>•What specific plans do you have for improvements in NDPP going forward?

I'll split this up in to the OpenMC tallying of NDPP data, and changes to NDPP itself.

OpenMC implementation of NDPP tallying:
* Aside from addressing comments in this PR, right now I am giving OpenMC the ability to create material-specific NDPP data.  This feature is 90% complete, and looks to significantly reduce the runtime penalty of using the NDPP data.  I haven't run my thesis sized problems (pin cell and 1/4 assembly-ish) yet, but have seen speedups of 150%.  I expect when I run those thesis problems, this material-specific tally should run at around the same speed as the collision based estimators (if not faster) which is right now only a 10% hit on runtime for a much larger improvement to uncertainty.  This comes with a memory decrement which really just comes down to the total number of nuclides vs the total number of materials in a problem.
* Another thing I would like to do is reduce the lines of code associated with this PR.  When doing the above material issue, I realized I would like to organize the code better such that the NDPP object itself knows how to produce its data given an incoming energy.  I can write this in such a way that I get rid of the generate_distrib_*n functions, and maybe even remove some of the tally_ndpp functions.
* Eventually it would be good to plug this in to CMFD so that, if available, one can use NDPP data for the scatter-1 tally used to create the diffusion coefficient.  I will probably need help testing this to make sure that the error associated with the NDPP data does not negatively affect the stability of the CMFD solution (I expect not due to the current-based corrections, but confirmation from someone smarter is a good thing).

NDPP Improvements:
* Right now NDPP was written to tackle 99% of the types of scattering distributions seen in ENDF VII.0 and VII.1 (see https://github.com/ndpp/ndpp/issues/17 for an example of how I wrote capability based on how ENDF represented some rare data).  I'd rather NDPP be re-factored to be more generic by nature so that it can handle any scattering reaction allowed by ACE directly, and if it can't, to stochastically integrate the distribution directly as a fall back.
* I would like to, in general re-factor NDPP to be more generic and approachable; I'd even be happy if I could convert most of the code base to python.
* The issues page (https://github.com/ndpp/ndpp/issues) shows some of the issues I've thought about so far.  The biggest are Issues 1,2, 4, and 5.
* Of course testing work will be needed with each and every ENDF release.


So, having said all the above, I will still keep working to address the comments brought up by @smharper and yourself, @paulromano since even if we do not incorporate this PR now, maintaining the coding practices brought up by the great reviews above will still be a net gain for OpenMC.

nelsonag commented:
All the above comments have been addressed and I did my  best to find other areas with the same formatting issues.  My next commit should resolve the failing tests.

nelsonag commented:
Welp, that didn't work either. Back to the travis_install.sh failure issue from last week.  This is really odd since the travis_install.sh apparently worked before my very last commit, but didnt when I removed a line from ndpp_lib.xml; of which travis_install should not care one bit about.

Of course, I can't reproduce this as travis_install and travis.sh work fine (and all pass w/in travis.sh) locally.  I did notice, however that travis_install.sh at times is using almost 4GB of RAM on my machine (every thing else is closed); it looks like we only have 3GB of RAM on Travis; is it possible the jobs are being killed? (that wouldn't explain why I that one line in ndpp_lib.xml matters of course)

wbinventor commented:
Although I've not had time to closely follow the discussion on this PR, I did want to chime in a bit with my thoughts on whether NDPP should be included in OpenMC, and whether it may find "organic" development from the open source community in the future.

**NDPP in OpenMC**
Personally, I think it would be cleanest to include NDPP as a submodule within OpenMC, unless there is good reason to think that another major MC code (*e.g.*, MCNP, Serpent) may pickup and use NDPP in the near to mid-term. My inclination is to think that this is not the case, and will not be the case unless more people begin using NDPP and publishing the results. However, I very may well be wrong about this. In any case, I think the workflow would be simplified if NDPP calculations were automatically dealt with at runtime by OpenMC.

**Sustainable NDPP Development**
Given that my PhD thesis is perhaps one of the most similar of any yet to @nelsonag's, I certainly have some interest in seeing that his NDPP creation benefit uses in the future. Although there may not be a large userbase for NDPP at the moment since few people use MC to generate MGXS for transport calculations, I think there is a good reason to believe that this may change in the future (and it will if I have anything to say about it). Although I very likely will not *need* NDPP for my PhD work, I do think that my work may make NDPP more practical and easier to use in the future through my (soon-to-be released) MGXS generation module for OpenMC. Of course this integration would be much simpler if NDPP were part of OpenMC, but I can very well see it being a loosely coupled as well. Since I don't really need NDPP, I likely won't be maintaining or extending NDPP itself, but do think there is a chance that the code for our two theses may be symbiotic and users may apply them together in the future.

nelsonag commented:
Thanks for the thoughts @wbinventor.  My only hesitation against packaging NDPP with OpenMC is that it  has such a large (typically one-time) run time overhead that we wouldn't want it to run as an initialization step in OpenMC.  So it should be a separate executable, and if a separate executable with different goals, why not a separate project?

Also, I am very excited for your MGXS module for OpenMC! Keep me updated!


p.s. The tests now pass, I re-ran it right after posting my last comment, and it magically worked.

nelsonag commented:
Any thoughts on the above @paulromano?

paulromano commented:
Yes, there are a few architectural things that need to be considered. Firstly, you may recall the [discussion on the openmc-dev list](https://groups.google.com/forum/#!topic/openmc-dev/LgtatEOxGuU) about a year ago regarding building up more nuclear data processing capabilities to lessen reliance on the ACE format. I do plan on pushing this effort forward this year, so it may be the case that NDPP makes more sense as part of that tool. My personal preference would be a Python-based tool with C, C++, or Cython under the hood for compute-intensive aspects, and it sounds like you might have the same thing in mind too. There are two likely side-effects:
- The data processing tool would start from ENDF files rather than ACE files. This means some of the NDPP integration routines would need to be reworked. You probably have a better sense than I how much of a pain in the ass that will be.
- I'd like the output to be HDF5-only to simplify both the tool itself as well as OpenMC. This means OpenMC would also rely solely on HDF5 for binary input/output.

I should mention that I've already started writing some Python scripts/modules for processing photoatomic and atomic relaxation data on my [photon branch](https://github.com/paulromano/openmc/tree/photon). That branch also has initial support for photon transport, which I'm hoping to submit a pull request for in the not-too-distant future.

So right now I'm leaning towards leaving your branch as a fork, and revisiting this when we have a more robust nuclear data processing capability which NDPP may be part of. Curious to hear what others think of all this. Is anyone opposed to moving towards an HDF5-only codebase?

nelsonag commented:
Everything you suggested as far as NDPP matches with where I would prefer this go as well.  I will continue to keep plugging away at my OpenMC branch as well as NDPP, and will close this pull request once it seems discussion re: your HDF5 question dies down.

I also have no objection towards the incorporation of NDPP functionality in to the data processing tool (which I see currently exists as openmc.processing, right?).  With PyNE, numpy, scipy, and some other potential python packages, an NDPP rewrite mostly in python could be quite straight-forward and hopefully not too many lines of code.  The utilization of ENDF data will require a little more work mostly due to handling some of the ENDF gotchas due to standard ambiguities (or even evaluation nuances).

For the record, I have no problem with moving towards an HDF5-only code base.

wbinventor commented:
I :heart: HDF5. And NYC.

smharper commented:
:+1: to HDF5

I think I would rather see NDPP (the tallying, not the data generation) pulled into OpenMC rather than left on it's own fork.  I think it gives us a unique and useful capability (I haven't tried it out on a real problem yet, but I certainly plan to).  And I don't think it detracts anything from OpenMC other than a little extra bulk in the code and documentation.  I don't think it's a maintenance issue; for example, I was able to convert it to the new `score_general` subroutine pretty easily.  I don't think it will interfere with the tally refactor.  In fact, it got me thinking about how we can make a tallying system that is more flexible overall (hopefully I share more on this topic later in the summer).

All that said, there is a non-trivial amount of work left for us in terms of testing and integrating it into the Python API.  I will defer to @nelsonag if he doesn't think it's worth the effort.

nelsonag commented:
Well, since the NDPP code itself is in stasis at the moment (as in it could undergo a large redesign as an integral data processing capability is incorporated into openmc) it probably is premature to include it in the main line. Having said that, I will still have my own branch (the same in this PR) that you could pull from.  If the uniqueness of the capability is something we want to put a feather in our cap for then we can move my branch to the main repo (as a branch distinct from develop and master) allowing it to be a more official developmental feature of openmc.

nelsonag commented:
Closing as the HDF5 discussion seems to be over.
